frame/support/src/storage/stream_iter.rs

```diff
@@ -0,0 +1,405 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{BoundedBTreeMap, BoundedBTreeSet, BoundedVec, WeakBoundedVec};
+use codec::Decode;
+use sp_std::vec::Vec;
+
+/// Provides the sealed trait `StreamIter`.
+mod private {
+	use super::*;
+
+	/// Used as marker trait for types that support stream iteration.
+	pub trait StreamIter {
+		/// The actual iterator implementation.
+		type Iterator: sp_std::iter::Iterator;
+
+		/// Create the stream iterator for the value found at `key`.
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator;
+	}
+
+	impl<T: codec::Decode> StreamIter for Vec<T> {
+		type Iterator = ScaleContainerStreamIter<T>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+
+	impl<T: codec::Decode> StreamIter for sp_std::collections::btree_set::BTreeSet<T> {
+		type Iterator = ScaleContainerStreamIter<T>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+
+	impl<K: codec::Decode, V: codec::Decode> StreamIter
+		for sp_std::collections::btree_map::BTreeMap<K, V>
+	{
+		type Iterator = ScaleContainerStreamIter<(K, V)>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+
+	impl<T: codec::Decode, S> StreamIter for BoundedVec<T, S> {
+		type Iterator = ScaleContainerStreamIter<T>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+
+	impl<T: codec::Decode, S> StreamIter for WeakBoundedVec<T, S> {
+		type Iterator = ScaleContainerStreamIter<T>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+
+	impl<K: codec::Decode, V: codec::Decode, S> StreamIter for BoundedBTreeMap<K, V, S> {
+		type Iterator = ScaleContainerStreamIter<(K, V)>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+
+	impl<T: codec::Decode, S> StreamIter for BoundedBTreeSet<T, S> {
+		type Iterator = ScaleContainerStreamIter<T>;
+
+		fn stream_iter(key: Vec<u8>) -> Self::Iterator {
+			ScaleContainerStreamIter::new(key)
+		}
+	}
+}
+
+/// An iterator that streams values directly from storage.
```

Comment:
We should also document that the value should not be modified as long as at least one Stream iter operates on it.